drivers/photo-electronic/adpd410x/iio_adpd410x.c

```diff
@@ -0,0 +1,228 @@
+
+	uint32_t freq = srt_to_uint32(buf);
+
+	ret = adpd410x_set_sampling_freq(dev, freq);
+	if (ret != SUCCESS)
+		return ret;
+
+	return len;
+}
+
+/**
+ * @brief Set number of active time slots.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t adpd410x_iio_set_last_timeslot(void *device, char *buf,
+		size_t len,
+		const struct iio_ch_info *channel, intptr_t priv)
+{
+	struct adpd410x_dev *dev = (struct adpd410x_dev *)device;
+	int32_t ret;
+
+	enum adpd410x_timeslots timeslot_no = srt_to_uint32(buf);
+
+	ret = adpd410x_set_last_timeslot(dev, timeslot_no);
+	if (ret != SUCCESS)
+		return ret;
+
+	return len;
+}
+
+/**
+ * @brief Set device sampling frequency.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t adpd410x_iio_set_opmode(void *device, char *buf, size_t len,
+				       const struct iio_ch_info *channel, intptr_t priv)
+{
+	struct adpd410x_dev *dev = (struct adpd410x_dev *)device;
+	int32_t ret;
+
+	enum adpd410x_opmode mode = srt_to_uint32(buf);
+
+	ret = adpd410x_set_opmode(dev, mode);
+	if (ret != SUCCESS)
+		return ret;
+
+	return len;
+}
+
+/** IIO channel attributes*/
+static struct iio_attribute adpd410x_iio_channel_attributes[] = {
+	{
+		.name = "raw",
+		.show = adpd410x_iio_channel_read_raw,
+		.store = NULL
+	},
+	END_ATTRIBUTES_ARRAY,
+};
+
+/** Channel Scan Type */
+static struct scan_type channel_scan_type = {
+	.sign = 'u',
+	.realbits = 24,
+	.storagebits = 32,
+	.shift = 0,
+	.is_big_endian = false
+};
+
+#define ADPD410X_IIO_CHANN_DEF(nm, ch) \
+	{ \
+		.name = nm, \
+		.ch_type = IIO_VOLTAGE, \
+		.channel = ch, \
+		.scan_type = &channel_scan_type, \
+		.attributes = adpd410x_iio_channel_attributes, \
+		.ch_out = false, \
+		.indexed = 1, \
+		.diferential = false, \
+	}
+
+/** IIO Channels */
+static struct iio_channel adpd410x_iio_channels[] = {
+	ADPD410X_IIO_CHANN_DEF("channel0", 0),
+	ADPD410X_IIO_CHANN_DEF("channel1", 1),
+	ADPD410X_IIO_CHANN_DEF("channel2", 2),
+	ADPD410X_IIO_CHANN_DEF("channel3", 3),
+	ADPD410X_IIO_CHANN_DEF("channel4", 4),
+	ADPD410X_IIO_CHANN_DEF("channel5", 5),
+	ADPD410X_IIO_CHANN_DEF("channel6", 6),
+	ADPD410X_IIO_CHANN_DEF("channel7", 7),
+	END_ATTRIBUTES_ARRAY
+};
+
+/** IIO attributes */
+static struct iio_attribute adpd410x_iio_attributes[] = {
+	{
+		.name = "sampling_frequency",
+		.show = NULL,
+		.store = adpd410x_iio_set_sampling_freq,
+	},
+	{
+		.name = "last_timeslot",
+		.show = NULL,
+		.store = adpd410x_iio_set_last_timeslot,
+	},
+	{
+		.name = "operation_mode",
+		.show = NULL,
```

Comment:
We could have a `show` method here too.

Reply:
wouldn't this imply having a getter inside the no-OS driver?
I would rather implement it there rather than in the IIO driver :)

Reply:
or a device attribute storing these values.

Reply:
I agree, we need it in the no-OS driver first. Whomever didn't include it in the first place was a fool.